src/main/kotlin/org/rust/ide/typing/paste/JsonStructParser.kt

```diff
@@ -140,12 +142,40 @@ private class StructParser {
         val innerType = when {
             foundDataTypes.size == 1 -> foundDataTypes.first()
             foundDataTypes.size == 2 && containsNull -> DataType.Nullable(foundDataTypes.first())
+            foundDataTypes.all { it is DataType.StructRef } -> unifyStructs(foundDataTypes.filterIsInstance<DataType.StructRef>())
             else -> DataType.Unknown
         }
 
         return DataType.Array(innerType)
     }
 
+    /**
+     * Look through a list of structs and try to unify them to a single struct type with optional fields.
+     */
+    private fun unifyStructs(structTypes: List<DataType.StructRef>): DataType {
+        val fieldMap = linkedMapOf<String, DataType>()
+        val fieldOccurrences = hashMapOf<String, Int>()
+
+        for (struct in structTypes) {
+            for (field in struct.struct.fields) {
+                fieldOccurrences[field.key] = fieldOccurrences.getOrDefault(field.key, 0) + 1
+                if (field.key !in fieldMap) {
```

Comment:
We should recursively unify nested structs if needed:
```json
{
  "items": [
    {
      "a": {
        "field1": 1
      }
    },
    {
      "a": {
        "field1": 1,
        "optional": 2
      }
    }
  ]
}
```

Reply:
Hmm, I was thinking about this and I didn't want to do it since it seems to be quite complex and I'm not sure if it's worth it.
Now the semantics are quite clear: go through the array, and if everything is an object, create a union of all found struct keys, and put values that were not present in all objects behind `Option`.

How should it work if it was recursive? We would have to collect all nested paths present in the contained objects, and each path that is not in all objects put it behind `Option`?
```
{
  "items": [
    {
      "a": {
        "field1":  {
             "c": { "x": 1 }
        }
      }
    },
    "a": {
     "field1": {"d": 2}
   }
  ]
}
```
Paths:
- `a.field1.c.x`
- `a.field1.d`

This should create `Option` field `d` and `Option` field `c`? What if get to a situation where not all of the items in a nested level will be objects?
```
[{"a": {
"b": 1
}}, {"a": {
"b": {"x": 5}
}}]
```

Reply:
I guess we can use something like this (where `generateType` was extracted from `parseArray`)
```kotlin
    private fun generateType(types: Set<DataType>): DataType {
        val containsNull = types.any { it is DataType.Nullable }
        val innerType = when {
            types.size == 1 -> types.first()
            types.size == 2 && containsNull -> DataType.Nullable(types.first())
            types.all { it is DataType.StructRef } -> unifyStructs(types.filterIsInstance<DataType.StructRef>())
            else -> DataType.Unknown
        }
        return innerType
    }

    /**
     * Look through a list of structs and try to unify them to a single struct type with optional fields.
     */
    private fun unifyStructs(structTypes: List<DataType.StructRef>): DataType {
        val fields = structTypes
            .flatMap { it.struct.fields.entries }
            .groupBy({ it.key }, { it.value })
            .mapValuesTo(linkedMapOf()) { generateType(it.value.toHashSet()) }

        val struct = Struct(fields)
        registerStruct(struct)
        return DataType.StructRef(struct)
    }
```

> What if get to a situation where not all of the items in a nested level will be objects?
In this case let's use Unknown type for now